common/experimental.yaml

```diff
@@ -12,3 +12,27 @@ datasets:
       reftype: object
     doc: Reference to the VectorData object that contains the enumerable elements
 
+datasets:
+- data_type_def: MeasurementData
+  data_type_inc: VectorData
+  dtype: numeric
+  doc: Data that are stored according to one dtype, but can be coerced by a
+    scaling factor and offset into another dtype representing the true units
+    of the measurement.
+  attributes:
+  - name: unit
+    dtype: text
+    doc: Base unit of measurement for working with the data.
+      Actual stored values are not necessarily stored in these units.
+      To access the data in these units, add the 'offset' to the 'data'
+      and multiply by the 'conversion'.
+  - name: offset
+    dtype: numeric
+      target_type: VectorData
+      reftype: object
+    doc: Factor that is added to shift the values of the stored data prior to scaling.
+  - name: conversion
+    dtype: numeric
+      target_type: VectorData
+      reftype: object
+    doc: Scalar multiple applied to all data shifted by offset.
```

Comment:
@oruebel Great point as well. One fewer operation where floating-point errors can be introduced.

I ran some performance tests for this and we really should stick with applying offsets first, and keep offsets in the same base `dtype` as the data: some tests showed large-scale `int` to `int` operations (add/subtract) seem to run 3-4 times faster than `int` to `float`.

Going to start moving ahead in a companion PR attempting to implement this in the code.

Reply:
My previous performance tests did not include cases where the result of the addition/subtraction by the offset requires an integer data type promotion. This came up in discussions with @alejoe91, who I'm including in this conversation to offer some advice since SpikeInterface has been dealing with this issue for years now.

For a clear example, `dtype=int8` (range of -128 to 127), `data=[127]`, and `offset=1` (`dtype` of `offset` does not matter). Adding the offset forces the `dtype` before conversion factor scaling to be bumped up to `int16` in order to return correct data values.

When re-running performance tests on somewhat large data arrays (~5GB buffers), cases requiring this `dtype` promotion do actually slow the data retrieval (with scaling) down by 20-30% compared to the method of applying the conversion factor first and thereafter working in `dtype=float`. Applying the offset first is only faster then if we can be absolutely confidence that the `data + offset` stays in the same `dtype`; unfortunately, I think the only way to do that super safely would be to scan the entire array for maximal values ahead of time, but that is an even more expensive operation. Further, the data scales required to incur significant floating-point errors from this additional operation are quite extreme so I wouldn't really be worried about that at all.

So actually, at this point, I would suggest doing what SpikeInterface has always done and what Ben originally suggested; `data_in_units = data_in_storage * conversion + offset` with everything `_in_units` being of `dtype=float` after the first conversion operation. This also has one nice interpretation - the offset value is itself given in the intended `unit` of the data and merely shifts it to around a central point.

We can definitely bring this up at our meeting next week (1/11/21) for further discussion, and I should have a dev branch on `hdmf` somewhat working by that point.